Sources/SwiftProtobuf/SwiftProtobufError.swift

```diff
@@ -0,0 +1,620 @@
+            self.storage.code = newValue
+        }
+    }
+
+    /// A message describing what went wrong and how it may be remedied.
+    public var message: String {
+        get { self.storage.message }
+        set {
+            self.ensureStorageIsUnique()
+            self.storage.message = newValue
+        }
+    }
+
+    public init(
+        code: Code,
+        message: String
+    ) {
+        self.storage = Storage(code: code, message: message)
+    }
+}
+
+extension SwiftProtobufError {
+    /// A high level indication of the kind of error being thrown.
+    public struct Code: Hashable, Sendable, CustomStringConvertible {
+        private enum Wrapped: Hashable, Sendable, CustomStringConvertible {
+            case binaryEncodingError
+            case binaryDecodingError
+            case jsonEncodingError
+            case jsonDecodingError
+            case textFormatDecodingError
+            case anyUnpackError
+            case protoFileToModuleMappingError
+
+            var description: String {
+                switch self {
+                case .binaryEncodingError:
+                    return "Binary encoding error"
+                case .binaryDecodingError:
+                    return "Binary decoding error"
+                case .jsonEncodingError:
+                    return "JSON encoding error"
+                case .jsonDecodingError:
+                    return "JSON decoding error"
+                case .textFormatDecodingError:
+                    return "Text format decoding error"
+                case .anyUnpackError:
+                    return "Error unpacking a Google_Protobuf_Any message"
+                case .protoFileToModuleMappingError:
+                    return "Error with proto file to module mapping"
+                }
+            }
+        }
+
+        public var description: String {
+            String(describing: self.code)
+        }
+
+        private var code: Wrapped
+        private init(_ code: Wrapped) {
+            self.code = code
+        }
+
+        public static var binaryEncodingError: Self {
+            Self(.binaryEncodingError)
+        }
+        
+        public static var binaryDecodingError: Self {
+            Self(.binaryDecodingError)
+        }
+        
+        public static var jsonEncodingError: Self {
+            Self(.jsonEncodingError)
+        }
+        
+        public static var jsonDecodingError: Self {
+            Self(.jsonDecodingError)
+        }
+        
+        public static var textFormatDecodingError: Self {
+            Self(.textFormatDecodingError)
+        }
+        
+        public static var anyUnpackError: Self {
+            Self(.anyUnpackError)
+        }
+        
+        public static var protoFileToModuleMappingError: Self {
+            Self(.protoFileToModuleMappingError)
+        }
+    }
+
+    /// A location within source code.
+    public struct SourceLocation: Sendable, Hashable {
```

Comment:
Do we want/need this?

For the different decoding errors, it seems like having support for a byteoffset and maybe line/column could someday be useful?

Reply:
We introduced this in almost all of our error structs now for debugging purposes so that you know where an error comes from. It has been really useful over time.

Reply:
However is this actually used somewhere @gjcairo ?

Reply:
Yes my bad, I was refactoring this and ended up being unused for the common errors I created. If we want to keep it I'll change all of the `static let`s into `func`s so we can pass in the location, otherwise I can get rid of this.

Reply:
> For the different decoding errors, it seems like having support for a byteoffset and maybe line/column could someday be useful?

@thomasvl Are you suggesting that the decoding-related `Code`s have the offset or line/column of the data that caused the error so that users can match on those, or is it fine if they're just added to the message string?

Reply:
I was thinking to the `Code`.  Generally I lean toward structured info vs. strings just because they tend to interoperate with other tooling better.  I'm also fine skipping it until we're actually going to provide the data.

Reply:
Gotcha - that sounds good to me. Maybe we can take care of that bit in a follow up PR? I'm aware this one's already quite large.

WRT the original question about the `SourceLocation`, @thomasvl do you think it's worth having this as part of the message description or would you rather not have this for some reason?

Reply:
> Gotcha - that sounds good to me. Maybe we can take care of that bit in a follow up PR? I'm aware this one's already quite large.

Makes sense.

> WRT the original question about the `SourceLocation`, @thomasvl do you think it's worth having this as part of the message description or would you rather not have this for some reason?

When you start to get into the space of client software, capturing `#file` can be risky as it allows someone to dump details out of the binary about the layout of the application's development. So I'm generally cautious of things like that.

Anyway, I'll have to defer to you all on this since it's your repo and the new Error interface should be what you all want.

Reply:
I think you bring up a very valid problem here with leaking file paths into potential logs or at worse user output. I just had to remind myself which magic identifier existed in the language today and I realised that we introduced the `#fileID` identifier exactly for this. @gjcairo can you make sure we rather use `#fileID` here since that only includes `ModuleName/FileName` instead of the full path.

Reply:
Okay I've added the location information to all common errors, and only expose it in the error description.